Implement m2m_to_o2m

carmenbianca opened the pull request:
Ported over from https://github.com/OCA/OpenUpgrade/pull/3911 in the hope that it can be more broadly useful.

Tagging @remytms

carmenbianca commented on openupgradelib/openupgrade.py:
```diff
@@ -1935,6 +1937,119 @@ def m2o_to_m2m(cr, model, table, field, source_field):
+    cr, source_relation_table, relation_source_field, relation_comodel_field
+):
+    """Warn user about data loss when migrating data from many2many to
+    many2one.
```
Good catch! Fixed.

pedrobaeza commented on openupgradelib/openupgrade.py:
```diff
@@ -1935,6 +1937,119 @@ def m2o_to_m2m(cr, model, table, field, source_field):
+    )
+
+
+def m2m_to_o2m_dataloss_warn(
```
Instead of 2 methods, put a keyword argument for warning.

carmenbianca replied:
good call, done.

pedrobaeza reviewed (commented):
Not sure about the use case, but this will probably needs a migration from one record with m2m to several records with m2o.

legalsylvain commented on openupgradelib/openupgrade.py:
```diff
@@ -1935,6 +1937,119 @@ def m2o_to_m2m(cr, model, table, field, source_field):
+    relation_comodel_field,
+):
+    """Transform many2many relations into one2many (with possible data
+    loss).
```
Could you explicit dataloss. With some exemple like : 
In VX : a1 related to B1, B2, B3. A2 related to B3, B4 , B5.
In VX+1 : ?

carmenbianca replied:
added better documentation + example.

carmenbianca commented:
Ready for review again. Changes since last update:

- Better documentation/naming. I decided to call the entities in the One2many relationship parent/child instead of source/comodel, which was not clear.
- Even more logging.
- Added a 'COPY' strategy that, instead of causing data loss, copies the child record for each parent record that wants one.

The copy strategy uses the Odoo ORM (`copy()`). I don't know if this is good/kosher/permissible. Doing this manually in SQL seemed like a Really Bad Idea™ to me.

carmenbianca commented:
Oof, syntax error because Travis uses Python 3.6.

legalsylvain commented:
> Oof, syntax error because Travis uses Python 3.6.
> edit: ImportError because Travis uses Python 3.6 :'(

We have to switch to github actions, but not done until now (#322)...

carmenbianca commented:
Turns out Python 2 was the problem, not Python 3.6. I added a backport dependency, but it wouldn't be very difficult to get rid of the Enum and use something more primitive.

MiquelRForgeFlow commented:
I don't think it's useful. No need the data loss. You need to duplicate records and link them. See https://github.com/OCA/OpenUpgrade/pull/4719.
